File: app/src/main/java/com/agrogames/islandsofwar/engine/impl/Engine.java
package com.agrogames.islandsofwar.engine.impl;

import android.os.Build;
import android.util.Log;

import androidx.annotation.RequiresApi;

import com.agrogames.islandsofwar.engine.abs.GameState;
import com.agrogames.islandsofwar.engine.abs.another.IGraphicsObject;
import com.agrogames.islandsofwar.engine.abs.common.Cell;
import com.agrogames.islandsofwar.engine.abs.bullet.IBullet;
import com.agrogames.islandsofwar.engine.abs.renderable.RenderableObject;
import com.agrogames.islandsofwar.engine.abs.unit.IUnit;
import com.agrogames.islandsofwar.engine.abs.map.MapObject;
import com.agrogames.islandsofwar.engine.impl.another.AnotherAdder;
import com.agrogames.islandsofwar.engine.impl.bullet.BulletAdder;
import com.agrogames.islandsofwar.engine.impl.map.MapProvider;
import com.agrogames.islandsofwar.engine.impl.unit.LandUnit;
import com.agrogames.islandsofwar.engine.impl.unit.UnitAdder;
import com.agrogames.islandsofwar.factories.Factory;
import com.agrogames.islandsofwar.map.impl.Water;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Engine implements com.agrogames.islandsofwar.engine.abs.Engine {
    private final List<IUnit> protectors = new ArrayList<>();
    private final List<IUnit> attackers = new ArrayList<>();
    private final List<IUnit> destroyed = new ArrayList<>();
    private final List<IBullet> protectorsBullets = new ArrayList<>();
    private final List<IBullet> attackersBullets = new ArrayList<>();
    private final List<IGraphicsObject> otherObjects = new ArrayList<>();
    private final MapObject[] mapObjects;
    private GameState state = GameState.Game;

    public Engine(IUnit[] protectors, IUnit[] attackers, MapObject[] mapObjects){
        this.protectors.addAll(Arrays.asList(protectors));
        this.attackers.addAll(Arrays.asList(attackers));
        this.mapObjects = mapObjects;
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public void update(float deltaTime) {
        updateObjects(deltaTime);
        deleteKilled();
        updateState();
    }

    private void updateObjects(float deltaTime) {
        IUnit[] protectorsCopy = protectors.toArray(new IUnit[0]);
        IUnit[] attackersCopy = attackers.toArray(new IUnit[0]);
        IBullet[] protectorsBulletsCopy = protectorsBullets.toArray(new IBullet[0]);
        IBullet[] attackersBulletsCopy = attackersBullets.toArray(new IBullet[0]);

        AnotherAdder anotherAdder = new AnotherAdder();
        MapProvider provider1 = new MapProvider(protectorsCopy, attackersCopy, mapObjects);
        BulletAdder bulletAdder1 = new BulletAdder();
        UnitAdder unitAdder1 = new UnitAdder();
        for (IUnit unit : protectorsCopy) {
            try{
                unit.update(provider1, bulletAdder1, unitAdder1, anotherAdder, deltaTime);
            }catch (Exception e){
                Log.e("IOW", "An exception has occurred during Update", e);
            }
        }
        for (IBullet bullet : protectorsBulletsCopy) {
            try{
                bullet.update(provider1, bulletAdder1, unitAdder1, anotherAdder, deltaTime);
            }catch (Exception e){
                Log.e("IOW", "An exception has occurred during Update", e);
            }
        }
        for(IUnit unit : destroyed.toArray(new IUnit[0])){
            unit.addTsd(deltaTime);
        }
        for (IGraphicsObject another : otherObjects.toArray(new IGraphicsObject[0])){
            another.update(provider1, bulletAdder1, unitAdder1, anotherAdder, deltaTime);
        }
        protectorsBullets.addAll(bulletAdder1.getBullets());
        protectors.addAll(unitAdder1.getUnits());

        BulletAdder bulletAdder2 = new BulletAdder();
        MapProvider provider2 = new MapProvider(attackersCopy, protectorsCopy, mapObjects);
        UnitAdder unitAdder2 = new UnitAdder();
        for (IUnit unit : attackersCopy) {
            try{
                unit.update(provider2, bulletAdder2, unitAdder2, anotherAdder, deltaTime);
            }catch (Exception e){
                Log.e("IOW", "An exception has occurred during Update", e);
            }
        }
        for (IBullet bullet : attackersBulletsCopy) {
            try{
                bullet.update(provider2, bulletAdder2, unitAdder2, anotherAdder, deltaTime);
            }catch (Exception e){
                Log.e("IOW", "An exception has occurred during Update", e);
            }
        }
        attackersBullets.addAll(bulletAdder2.getBullets());
        attackers.addAll(unitAdder2.getUnits());

        otherObjects.addAll(anotherAdder.getAnotherObjects());
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    private void deleteKilled() {
        for (IUnit unit : protectors.toArray(new IUnit[0])) {
            if(unit.getHealth().current <= 0){
                if(unit instanceof LandUnit){
                    protectors.remove(unit);
                    otherObjects.add(Factory.getGraphics("bang", unit.getLocation(), unit.getRotation()));
                } else {
                    if(!destroyed.contains(unit)) destroyed.add(unit);
                }
            }
        }
        for (IUnit unit : attackers.toArray(new IUnit[0])) {
            if(unit.getHealth().current <= 0){
                if(unit instanceof LandUnit){
                    attackers.remove(unit);
                    otherObjects.add(Factory.getGraphics("bang", unit.getLocation(), unit.getRotation()));
                } else {
                    if(!destroyed.contains(unit)) destroyed.add(unit);
                }
            }
        }
        for(IUnit unit : destroyed.toArray(new IUnit[0])){
            if(unit.timeSinceDestroyed() >= 3.5f){
                destroyed.remove(unit);
                attackers.remove(unit);
                if(Arrays.stream(mapObjects).noneMatch(c -> c instanceof Water && Arrays.asList(c.getTerritory()).contains(new Cell(unit.getLocation())))){
                    otherObjects.add(Factory.getGraphics("big_bang", unit.getLocation(), unit.getRotation()));
                }
                for (Cell cell : unit.getTerritory()){
                    for(IUnit u : protectors.stream().filter(c -> Arrays.asList(c.getTerritory()).contains(cell)).toArray(IUnit[]::new)){
                        u.loseHealth(15);
                    }
                    for(IUnit u : attackers.stream().filter(c -> Arrays.asList(c.getTerritory()).contains(cell)).toArray(IUnit[]::new)){
                        u.loseHealth(15);
                    }
                }
            }
        }
        for(IGraphicsObject graphicsObject : otherObjects.toArray(new IGraphicsObject[0])){
            if(graphicsObject.shouldBeRemoved()){
                otherObjects.remove(graphicsObject);
            }
        }

        for (IBullet bullet : protectorsBullets.toArray(new IBullet[0])) {
            if(bullet.hasStopped()){
                protectorsBullets.remove(bullet);
            }
        }
        for (IBullet bullet : attackersBullets.toArray(new IBullet[0])) {
            if(bullet.hasStopped()){
                attackersBullets.remove(bullet);
            }
        }
    }

    private void updateState(){
        if(state != GameState.Game) return;
        if(attackers.size() == 0) state = GameState.Defeat;
        else if(protectors.size() == 0) state = GameState.Win;
    }

    public IUnit[] getProtectors(){
        return protectors.toArray(new IUnit[0]);
    }

    public IUnit[] getAttackers(){
        return attackers.toArray(new IUnit[0]);
    }

    @Override
    @RequiresApi(Build.VERSION_CODES.N)
    public RenderableObject[] getOther() {
        List<RenderableObject> objects = new ArrayList<>();
        objects.addAll(destroyed);
        objects.addAll(otherObjects);
        return objects.toArray(new RenderableObject[0]);
    }

    public IBullet[] getProtectorsBullets(){
        return protectorsBullets.toArray(new IBullet[0]);
    }

    public IBullet[] getAttackersBullets(){
        return attackersBullets.toArray(new IBullet[0]);
    }

    @Override
    public GameState getState() {
        return state;
    }

    @Override
    public void addPlane(IUnit plane) {
        attackers.add(plane);
    }
}

File: app/src/main/java/ru/agrogames/islands/engine/impl/Engine.kt
package ru.agrogames.islands.engine.impl

import android.util.Log
import ru.agrogames.islands.engine.abs.GameState
import ru.agrogames.islands.engine.abs.bullet.IBullet
import ru.agrogames.islands.engine.abs.common.Cell
import ru.agrogames.islands.engine.abs.graphics.IGraphicsObject
import ru.agrogames.islands.engine.abs.map.MapObject
import ru.agrogames.islands.engine.abs.renderable.RenderableObject
import ru.agrogames.islands.engine.abs.unit.IUnit
import ru.agrogames.islands.engine.impl.bullet.BulletAdder
import ru.agrogames.islands.engine.impl.graphics.GraphicsAdder
import ru.agrogames.islands.engine.impl.map.MapProvider
import ru.agrogames.islands.engine.impl.unit.LandUnit
import ru.agrogames.islands.engine.impl.unit.UnitAdder
import ru.agrogames.islands.factories.Factory
import ru.agrogames.islands.islands.abs.Island
import ru.agrogames.islands.map.Water
import java.util.LinkedList

class Engine(protectors: List<IUnit>, attackers: List<IUnit>, private val mapObjects: Array<MapObject>, val mapName: String) {
    val protectors: MutableList<IUnit> = LinkedList(protectors)
    val attackers: MutableList<IUnit> = LinkedList(attackers)
    private val destroyed: MutableList<IUnit> = LinkedList()
    val protectorsBullets: MutableList<IBullet> = LinkedList()
    val attackersBullets: MutableList<IBullet> = LinkedList()
    private val otherObjects: MutableList<IGraphicsObject> = LinkedList()
    var state = GameState.Game
        private set

    constructor(island: Island) : this(listOf(*island.owners.clone()), listOf(island.attacker), island.map.map, island.map.name)
    fun update(deltaTime: Float) {
        try {
            updateObjects(deltaTime)
            deleteKilled()
            updateState()
        } catch (e: Exception){
            Log.e("IOW", "ERROR IN UPDATE", e)
        }
    }

    private fun updateObjects(deltaTime: Float) {
        val protectorsCopy = protectors.toTypedArray()
        val attackersCopy = attackers.toTypedArray()
        val protectorsBulletsCopy = protectorsBullets.toTypedArray()
        val attackersBulletsCopy = attackersBullets.toTypedArray()
        val graphicsAdder = GraphicsAdder()
        val provider1 = MapProvider(protectorsCopy, attackersCopy, mapObjects)
        val bulletAdder1 = BulletAdder()
        val unitAdder1 = UnitAdder()
        for (unit in protectorsCopy) {
            try {
                unit.update(provider1, bulletAdder1, unitAdder1, graphicsAdder, deltaTime)
            } catch (e: Exception) {
                Log.e("IOW", "An exception has occurred during Update", e)
            }
        }
        for (bullet in protectorsBulletsCopy) {
            try {
                bullet.update(provider1, bulletAdder1, unitAdder1, graphicsAdder, deltaTime)
            } catch (e: Exception) {
                Log.e("IOW", "An exception has occurred during Update", e)
            }
        }
        for (unit in destroyed.toTypedArray()) {
            unit.addTsd(deltaTime)
        }
        for (another in otherObjects.toTypedArray<IGraphicsObject?>()) {
            another!!.update(provider1, bulletAdder1, unitAdder1, graphicsAdder, deltaTime)
        }
        protectorsBullets.addAll(bulletAdder1.bullets)
        protectors.addAll(unitAdder1.units)
        val bulletAdder2 = BulletAdder()
        val provider2 = MapProvider(attackersCopy, protectorsCopy, mapObjects)
        val unitAdder2 = UnitAdder()
        for (unit in attackersCopy) {
            try {
                unit.update(provider2, bulletAdder2, unitAdder2, graphicsAdder, deltaTime)
            } catch (e: Exception) {
                Log.e("IOW", "An exception has occurred during Update", e)
            }
        }
        for (bullet in attackersBulletsCopy) {
            try {
                bullet.update(provider2, bulletAdder2, unitAdder2, graphicsAdder, deltaTime)
            } catch (e: Exception) {
                Log.e("IOW", "An exception has occurred during Update", e)
            }
        }
        attackersBullets.addAll(bulletAdder2.bullets)
        attackers.addAll(unitAdder2.units)
        otherObjects.addAll(graphicsAdder.graphicsObjects)
    }

    private fun deleteKilled() {
        for (unit in protectors.toTypedArray()) {
            if (unit.health.current <= 0) {
                if (unit is LandUnit) {
                    protectors.remove(unit)
                    otherObjects.add(Factory.getGraphics("bang", unit.location, unit.rotation))
                } else {
                    if (!destroyed.contains(unit)) destroyed.add(unit)
                }
            }
        }
        for (unit in attackers.toTypedArray()) {
            if (unit.health.current <= 0) {
                if (unit is LandUnit) {
                    attackers.remove(unit)
                    otherObjects.add(Factory.getGraphics("bang", unit.location, unit.rotation))
                } else {
                    if (!destroyed.contains(unit)) destroyed.add(unit)
                }
            }
        }
        for (unit in destroyed) {
            if (unit.timeSinceDestroyed() >= 3.5f) {
                destroyed.remove(unit)
                attackers.remove(unit)
                if (mapObjects.none { it is Water && it.territory.contains(Cell(unit.location)) }) {
                    otherObjects.add(Factory.getGraphics("big_bang", unit.location, unit.rotation))
                }
                for (cell in unit.territory) {
                    for (u in protectors.filter { it.territory.contains(cell) }) {
                        u.loseHealth(15)
                    }
                    for (u in attackers.filter { it.territory.contains(cell) }) {
                        u.loseHealth(15)
                    }
                }
            }
        }
        for (graphicsObject in otherObjects.toTypedArray()) {
            if (graphicsObject.shouldBeRemoved()) {
                otherObjects.remove(graphicsObject)
            }
        }
        for (bullet in protectorsBullets.toTypedArray()) {
            if (bullet.hasStopped()) {
                protectorsBullets.remove(bullet)
            }
        }
        for (bullet in attackersBullets.toTypedArray()) {
            if (bullet.hasStopped()) {
                attackersBullets.remove(bullet)
            }
        }
    }

    private fun updateState() {
        if (state != GameState.Game) return
        if (attackers.size == 0) state = GameState.Defeat else if (protectors.size == 0) state = GameState.Win
    }

    val other: List<RenderableObject>
        get() {
            val objects: MutableList<RenderableObject> = ArrayList()
            objects.addAll(destroyed)
            objects.addAll(otherObjects)
            return objects
        }


    fun addPlane(plane: IUnit) {
        attackers.add(plane)
    }
}